src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -1110,18 +1110,26 @@ private void HandleReadResponseContentException(Exception ex, CancellationToken
                     throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, abortException));
                 case Http3ConnectionException _:
                     // A connection-level protocol error has occurred on our stream.
-                    _connection.Abort(ex);
+                    _connection?.Abort(ex);
                     throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
                 case OperationCanceledException oce when oce.CancellationToken == cancellationToken:
-                    _stream.AbortWrite((long)Http3ErrorCode.RequestCancelled);
+                    _stream?.AbortRead((long)Http3ErrorCode.RequestCancelled);
                     ExceptionDispatchInfo.Throw(ex); // Rethrow.
                     return; // Never reached.
                 default:
-                    _stream.AbortWrite((long)Http3ErrorCode.InternalError);
+                    _stream?.AbortRead((long)Http3ErrorCode.InternalError);
                     throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
             }
         }
 
+        private void CancelResponseContentRead()
+        {
+            if (Volatile.Read(ref _responseDataPayloadRemaining) != -1) // -1 indicates EOS
```

Comment:
There is an inherent race between setting this and checking it in ReadNextDataFrameAsync.

That means you need to handle the case where the AbortRead happens after we check this in ReadNextDataFrameAsync, but before we try to actually read the frame. 

As such, I would suggest not even trying to set `_responseDataPayloadRemaining` here. Just do AbortRead and let the next attempt to read from the underlying stream fail, and handle it as appropriate.